Register and get user

gberaudo opened the pull request:
Register a user.
Get user info by id.

gberaudo commented:
> Should we use a more expressive route? E.g. POST /users/register instead of `POST /users'
@tsauerwein, indeed it would be greater. Do I have to add a new `resource` for this? Or switch to the `service` API?

tsauerwein commented:
> Do I have to add a newresourcefor this? Or switch to theservice` API?

I think a 2nd resource with the path `collection_path='/users/register'` could do the trick.

gberaudo commented:
> Why is it necessary to get the password in utf-8? Isn't everything already in utf-8?
No idea, I get `Unicode-objects must be encoded before hashing` error otherwise.

gberaudo commented:
Ready for a new review.

gberaudo commented:
Almost. Password requirement does not work though it is in the schema. Any idea why?

tsauerwein commented:
> Password requirement does not work though it is in the schema. Any idea why?

Probably because `password` is a property, which might not be understood by ColanderAlchemy.

tsauerwein commented on c2corg_api/views/user.py:
```diff
@@ -0,0 +1,65 @@
+        except:
+            # Catching the exception ensures a json is sent back to the user
+            # and a message is shown.
+            self.request.errors.add('user', 'error', 'an error occured')
```
The three values for the error have a specific meaning, see:
http://cornice.readthedocs.org/en/latest/validation.html#dealing-with-errors

You could use:

    self.request.errors.add('body', 'user', 'error creating user')

tsauerwein replied:
But I am wondering: If you simply add something to this `errors` list (outside the validation), does this return the error messages? Because one line below you are returning the user as if everything worked fine.
I think you would have to raise a `BadRequest` like here: https://github.com/c2corg/v6_api/blob/master/c2corg_api/views/document.py#L321-L322

gberaudo replied:
Yes, adding something to the list will return validation errors.
But you are right raising would make sense; it is client code duty to detect HTTP errors and update interface accordingly. I think in this case an `InternalError` would be great.

Shouldn't we log the error? I see we are not using loggers yet.
How do we deal with sensitive data?

tsauerwein replied:
> Yes, adding something to the list will return validation errors.

Inside a validation function, yes. But outside I think no?

> Shouldn't we log the error? I see we are not using loggers yet.

Yes, I created a ticket: https://github.com/c2corg/v6_api/issues/78

> How do we deal with sensitive data?

What do you mean? (Never log a password or token, but I think that is clear.)

gberaudo replied:
I mean that some stack traces may contain sensitive data.

tsauerwein commented:
You could explicitly add the password field to the schema:

    from colander import SchemaNode, String as ColanderString, required
    ...
    schema_create_user.children.append(SchemaNode(ColanderString(), name='password', missing=required))

(untested, but something in this direction)

gberaudo commented:
I fixed the validation issue with `password`. I tried something similar to what you propose which I found
in the colander tests code but it did not work, so I am simply using a validator.
In some way it makes sense because we only handle deserialization.

gberaudo commented:
@asaunier, do you know if users can use non-ascii characters in v5?
I currently accept unicode as password input but it could easily be restricted to ascii instead.

gberaudo commented:
I think it is now ready for merging. Do you have more comments?

asaunier commented:
> @asaunier, do you know if users can use non-ascii characters in v5?

You mean for passwords?
I think that any UTF-8 character is accepted. Why restricting to ascii?

The code of v5 about password updating:
https://github.com/c2corg/camptocamp.org/blob/602b46c9388436cbbd21f99dde856065d33d0fed/apps/frontend/modules/users/actions/actions.class.php#L458-L470

When someone signs in, it seems a password is auto-generated and send by email (same tool than for get a new password when you don't remember your password).

Generated passwords actually use ascii characters:
https://github.com/c2corg/camptocamp.org/blob/602b46c9388436cbbd21f99dde856065d33d0fed/lib/model/doctrine/UserPrivateData.class.php#L64-L80
but I have seen no restriction (?) when the user updates his password.

If really need, restricting passwords to ascii characters could be possible. It means that users we non-ascii passwords would have to ask for a new password. That would be better to avoid this but that's a possibility.

asaunier commented on c2corg_api/views/user.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def validate_json_password(request):
+    """Checks if the password was given and encode it.
```
encodeS (or remove the trailing s in Checks)

asaunier commented on c2corg_api/views/user.py:
```diff
@@ -0,0 +1,86 @@
+        request.errors.add('user', 'password', 'Required')
+
+    try:
+        # We receives an unicode string. The hashing function used
```
an unicode => a unicode

asaunier commented:
I have no other comment.

gberaudo commented:
Fixed the typos, thanks @asaunier.
Regarding passwords, notably utf8, I added an item in https://github.com/c2corg/v6_api/issues/70 so that it is correctly checked.

gberaudo commented:
Thanks for the reviews @tsauerwein and @asaunier.
